python/xorbits/core/data.py

```diff
@@ -212,6 +212,23 @@ def gen():
 
                 yield from gen()
 
+    def __len__(self):
+        from .._mars.core import HasShapeTileable
+        from .execution import run
+
+        if isinstance(self.data._mars_entity, HasShapeTileable):
+            try:
+                return int(self.data._mars_entity.shape[0])
+            except (IndexError, ValueError):
+                # shape is unknown, execute it
+                run(self)
+                try:
+                    return int(self.data._mars_entity.shape[0])
+                except (IndexError, ValueError):  # happens when dimension is 0
+                    return 0
+        else:
+            raise TypeError(f"object of type '{self.data.data_type}' has no len()")
```

Comment:
The magic method wrapper raises attribute errors when the underlying object does not have a specify magic method:
```python
raise AttributeError(
                f"'{self.data.data_type.name}' object has no attribute '{method_name}'"
            )
```

Would it be better to also raise an attribute error here?

Reply:
This is Python's behavior:
``` Python
In [1]: class A: pass

In [2]: len(A())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 len(A())

TypeError: object of type 'A' has no len()
```